### ISSUE TITLE
JCryptCipher - Chosen Ciphertext Attacks

### ISSUE THREAD

**Issue by paragonie-scott** [joomla/joomla-cms#8329]
None of the JCrypt implementations are properly [authenticating the ciphertext](https://paragonie.com/blog/2015/05/using-encryption-and-authentication-correctly). As a result, an active attacker can flip bits at will until they get a useful result. Combine the CCA with #8328 and you get a nice padding oracle attack against the Rijndael256-CBC encryption.

Further reading: 

* http://web.engr.oregonstate.edu/~rosulekm/crypto/12-cca.pdf
* http://tonyarcieri.com/all-the-crypto-code-youve-ever-written-is-probably-broken
* http://crypto.stackexchange.com/questions/1263/how-does-a-chosen-ciphertext-attack-work-with-a-simple-example

TL;DR Encrypt then MAC. See how [defuse/php-encryption](https://github.com/defuse/php-encryption) implements its encryption and authentication. (In fact, I highly recommend switching to that and exercising liberal use of your delete key on the existing cryptography code.)

**Comment by paragonie-scott** [comment]
Update: It appears that the IV is a static value rather than a per-message random value. This sort of defeats the entire purpose of using CBC mode.

**Comment by Bakual** [comment]
`JCryptCipherSimple` is already deprecated and will be removed with 4.0.
Due to B/C it is still shiped but it's recommended to use the newer bcrypt based encryption.

We changed our encryption for passwords (and other things) to bcrypt with Joomla 3.3.0 which has been released 2014-04-30 and raised minimum PHP requirements to 5.3.10 due to this security change.

May I ask you to send potential security issues in future to our security team using either https://developer.joomla.org/contact-security-team.html or security@joomla.org.

**Comment by PhilETaylor** [comment]
@paragonie-scott Would you be able to provide a mergeable pull request with the changes you are proposing please? Code speaks louder than comments :-)

**Comment by paragonie-scott** [comment]
You won't like my pull request here.

**Comment by mbabker** [comment]
Unfortunately the php-encryption library requires PHP 5.4 while we're still in 5.3.10+ world (ya, shoot us ;-) ).  So that's not much of an option.

**Comment by mbabker** [comment]
@joomla/security Some help on this issue would be appreciated.  I'm in over my head on this.

**Comment by SniperSister** [comment]
I only have very limited knowledge on crypto, not really sure how to help with this :(

**Comment by paragonie-scott** [comment]
You could reference what [defuse/php-encryption v1.1](https://github.com/defuse/php-encryption/tree/v1.1) did (or just use that outright). It uses mcrypt and supports PHP 5.3.

To prevent a chosen ciphertext attack:

1. Split your key into two keys: `$eKey` and `$aKey`. HKDF-SHA-256 can do this safely (Defuse's library includes an implementation).
2. Pad your plaintext using PKCS7.
3. Encrypt your data with `$eKey` (and preferably with a 16-byte random IV from `random_bytes()`, provided by random_compat).
4. Take the ciphertext (and the random IV) from step 3, authenticate the entire string with HMAC-SHA-256, and append that to the message.
5. (Optional) base64 or hex encode the output for safe storage / transport.

When decrypting an encrypted message:

1. (Optional) Decode back to raw binary. Validate that nothing failed.
2. Split the master key into `$eKey` and `$aKey` again.
3. Recalculate the MAC of the given IV and ciphertext, using `$aKey`.
4. Compare (using `hash_equals()`) that the MAC provided with the message is equal to the one you just recalcualted. If it fails, throw an exception. (If the user fails to handle the exception, it *must* terminate execution.)
5. Decrypt the ciphertext using the IV and `$eKey`.
6. Remove the PKCS7 padding.

This also resolves the data corruption by `trim()` issue. But seriously, use the version 1.1 tag of defuse/php-encryption instead of writing your own. When you can support modern PHP (5.6+ or, by the time that comes up, likely 7.x+), you can just  change the version requirements for Defuse's library to 1.2 or 2.0.

cc @defuse who can agree/disagree with whether using 1.1 of his library is a good idea or not. (I rewrote it to leverage OpenSSL for 1.2, which introduced the 5.4+ requirement.)

**Comment by paragonie-scott** [comment]
That said, PR incomin'

**Comment by Bakual** [comment]
> That said, PR incomin

Awesome, thanks a lot!

**Comment by paragonie-scott** [comment]
I think we can remove the no-code-attached label now ;)

**Comment by wilsonge** [comment]
I'm going to close this issue so that the discussion can continue on the Pull Request :)
